# Project 3: Confidence Intervals in Astronomical Data

## Research Question 
What is the average redshift of quasars in our SDSS sample,
and how precisely can we estimate the population mean?

## Introduction
In this project, I will use confidence intervals and bootstrap
resampling to quantify uncertainty in estimates from astronomical data. By the end of this project, I want to learn the basics of statistics, confidence intervals, random sampling, visualizing uncertainty, and interpreting statistical results in an astronomy context.

First, we import libraries and download the data set. Run the cell below.

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

sdss17 = pd.read_csv("../data/star_classification.csv")

For a refresher on how the data looks like, run the cell below.

In [5]:
print(sdss17.head())
sdss17.info()

         obj_ID       alpha      delta         u         g         r  \
0  1.237661e+18  135.689107  32.494632  23.87882  22.27530  20.39501   
1  1.237665e+18  144.826101  31.274185  24.77759  22.83188  22.58444   
2  1.237661e+18  142.188790  35.582444  25.26307  22.66389  20.60976   
3  1.237663e+18  338.741038  -0.402828  22.13682  23.77656  21.61162   
4  1.237680e+18  345.282593  21.183866  19.43718  17.58028  16.49747   

          i         z  run_ID  rerun_ID  cam_col  field_ID   spec_obj_ID  \
0  19.16573  18.79371    3606       301        2        79  6.543777e+18   
1  21.16812  21.61427    4518       301        5       119  1.176014e+19   
2  19.34857  18.94827    3606       301        2       120  5.152200e+18   
3  20.50454  19.25010    4192       301        3       214  1.030107e+19   
4  15.97711  15.54461    8102       301        3       137  6.891865e+18   

    class  redshift  plate    MJD  fiber_ID  
0  GALAXY  0.634794   5812  56354       171  
1  GALAXY  0.77913

## Setting Up
Our research question concerns quasars, so we will need to create a DataFrame containing only quasars using the filtering methods we learned.

In [9]:
quasars = sdss17[sdss17['class'] == 'QSO']
quasars.head()

,obj_ID,alpha,delta,u,g,r,i,z,run_ID,rerun_ID,cam_col,field_ID,spec_obj_ID,class,redshift,plate,MJD,fiber_ID
5,1.237680e+18,340.995121,20.589476,23.48827,23.33776,21.32195,20.25615,19.54544,8102,301,3,110,5.658977e+18,QSO,1.424659,5026,55855,741
6,1.237679e+18,23.234926,11.418188,21.46973,21.17624,20.92829,20.60826,20.42573,7773,301,2,462,1.246262e+19,QSO,0.586455,11069,58456,113
17,1.237679e+18,1.494389,3.291746,20.38562,20.40514,20.29996,20.05918,19.89044,7712,301,5,339,9.843824e+18,QSO,2.031528,8743,57663,295
23,1.237658e+18,145.883006,47.300484,21.73992,21.53095,21.26763,21.36257,21.15861,2821,301,2,33,8.228240e+18,QSO,2.075680,7308,56709,596
33,1.237679e+18,20.052556,11.497881,21.89214,21.35124,21.18755,20.84300,20.76580,7773,301,2,441,1.246830e+19,QSO,1.528308,11074,58488,306


In [11]:
quasars.shape

(18961, 18)

#### Question 1
What is our sample?

Our sample are the quasars in the SDSS17 data set.

#### Question 2
What could we consider the population?

The population is all of the celestial observations from the SDSS17 data set.

#### Question 3
What variable are we studying?

We are studying the redshift of quasars in the SDSS17 data set.

###
Now we know what our research question is about, we can narrow the data down even further

In [14]:
redshift = quasars['redshift']
redshift.describe()

count    18961.000000
mean         1.719676
std          0.913954
min          0.000461
25%          1.106605
50%          1.617232
75%          2.220279
max          7.011245
Name: redshift, dtype: float64

In [15]:
redshift.head()

5     1.424659
6     0.586455
17    2.031528
23    2.075680
33    1.528308
Name: redshift, dtype: float64

In [16]:
redshift.isna().sum()

np.int64(0)

isna() is a function that counts how many missing redshifts are in our quasar sample. Because we have 0, we don't have to do anything else. In the case that the output was greater than 0, we would deal with them using dropna() which drops all the missing values.

## Our First Statistics

The mean is the mathematical average set of numbers. 

In [17]:
redshift.mean()

np.float64(1.7196760892911345)

The standard deviation is a number that shows how spread out the values in a data set are from the average. The smaller the standard deviation, the tighter the points are spread across and are closer to the mean. This means the values are steady, reliable, and similar. A high standard deviation means the data points are spread far and show high variation. A zero standard deviation means there's no variation at all, and the numbers in the data set are identical.

In [18]:
redshift.std()

np.float64(0.9139542763221472)

In [19]:
len(redshift)

18961

Interpretation:  Our mean of about 1.72 represents the average redshift of quasars. Our quasar redshifts typically vary by about 0.914 redshift units around the sample mean's redshift, indicating a relatively wide spread around the mean. Our sample contains 18,961 quasar redshift observations. This provides a relatively large sample for estimating the average redshift of quasars in our dataset.